bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py

```diff
@@ -262,6 +267,291 @@ def predict(self):
+                for j in range(params.classes):
+                    class_index = self._entry_index(params.side, params.coords, params.classes, n * side_square + i,
+                                            params.coords + 1 + j)
+                    confidence = scale * predictions[class_index]
+                    if confidence < threshold:
+                        continue
+                    objects.append(self._scale_bbox(x=x, y=y, h=h, w=w, class_id=j, confidence=confidence,
+                                            h_scale=orig_im_h, w_scale=orig_im_w))
+        return objects
+
+    def _intersection_over_union(self, box_1, box_2):
+        """
+        Calculates an iou of two bounding boxes.
+
+        :param box_1: first bounding box
+        :param box_2: second bounding box
+        :return: iou value
+        """
+        # Calculate overlap
+        width_of_overlap_area = min(box_1['xmax'], box_2['xmax']) - max(box_1['xmin'], box_2['xmin'])
+        height_of_overlap_area = min(box_1['ymax'], box_2['ymax']) - max(box_1['ymin'], box_2['ymin'])
+        if width_of_overlap_area < 0 or height_of_overlap_area < 0:
+            area_of_overlap = 0
+        else:
+            area_of_overlap = width_of_overlap_area * height_of_overlap_area
+        # Calculate areas
+        box_1_area = (box_1['ymax'] - box_1['ymin']) * (box_1['xmax'] - box_1['xmin'])
+        box_2_area = (box_2['ymax'] - box_2['ymin']) * (box_2['xmax'] - box_2['xmin'])
+        # Calculate union
+        area_of_union = box_1_area + box_2_area - area_of_overlap
+        # Catch divide by zero
+        if area_of_union == 0:
+            return 0
+        # Return iou
+        return area_of_overlap / area_of_union
+
+    def _scaled_bbox_to_candidate(self, bbox):
+        return Candidate(
+            int(bbox['xmin']),
+            int(bbox['ymin']),
+            int(bbox['xmax']) - int(bbox['xmin']),
+            int(bbox['ymax']) - int(bbox['ymin']),
+            bbox['confidence']
+            )
+
+    def predict(self):
+        if (self._ball_candidates is None and self._goalpost_candidates is None) or not self._caching:
+            # Set up variables
+            self._ball_candidates = list()
+            self._goalpost_candidates = list()
+
+            rospy.logdebug("Starting inference...")
+
+            # Set request id
+            request_id = 1
+            # Resize image to yolo input size
+            in_frame = cv2.resize(self._image, (self._w, self._h))
+
+            # resize input_frame to network size
+            in_frame = in_frame.transpose((2, 0, 1))  # Change data layout from HWC to CHW
+            in_frame = in_frame.reshape((self._n, self._c, self._h, self._w))
+
+            # Start inference
+            self._exec_net.start_async(request_id=request_id, inputs={self._input_blob: in_frame})
+
+            # Collecting object detection results
+            objects = list()
+            # Create barrier
```

Comment:
```suggestion
            # Create barrier to check only one call is made to the net at any given time
```
just explaining what is meant by barrier

Reply:
But that's not the point of the barrier. We only want to wait until the net finished its prediction, to move on with the next precessing steps.

Reply:
```suggestion
            # Create barrier. This lets all following processing steps wait until the prediction is calculated.
```